### PR TITLE
New options: writing tables directly into a .root file and MC tracks filtering.

### PR DESCRIPTION
**Pull request by nburmaso**
*Writing tables directly into a .root file using the new option `--ignore-aod-writer 1`
*MC particles can now be filtered using `--reco-mctracks-only 1` to store only reconstructed tracks and all their "relatives" up to primary particles. Primary particles will be saved even if they are not reconstructed. For correct functioning user needs either ITS tracks or TPC tracks to be stored (or both).

*Do not merge yet, the features need to be tested.

### CODE REVIEW

**Comment by jgrosseo** [comment]
Why are we skipping the AOD writer?

**Comment by nburmaso** [comment]
Hi @jgrosseo. AOD writer is skipped in order to free memory as soon as the tables are filled. It is yet to be implemented and I still have some problems, but I decided to share the code to show the current state of work.

**Comment by jgrosseo** [comment]
OK, thanks. Could you create a jira ticket about the problem, so that we don't forget about it?

**Comment by nburmaso** [comment]
Sure, I will make a ticket for this

**Comment by sawenzel** [comment]
For the record, there is a JIRA for this initiative https://alice.its.cern.ch/jira/browse/O2-2091.

I suggested to have an option to skip AOD writer because in the usual setting of split producer/writer we are forced to accumulate all tables in memory within the run-function invocation (and don't have control YET to release memory). For a 100PbPb timeframe, I measured the max mem consumption to be ~40GB which is substantially above acceptable limits (in particular since this is essentially a 1-core process).

By allowing to write the AOD directly in the producer, we would be able to manage memory ourselves and alloc/free it on a much more granular basis.

**Comment by jgrosseo** [comment]
Do we understand why the writer does not free the tables? I had the impression that in the analysis filtering this is done...

**Comment by sawenzel** [comment]
The writer may well do it (when it shuts down) but then it's too late. For me the problem is really the sequence: a) We accumulate N tables, then we write N tables, then we free. Rather than the extreme limit b) We build 1 table, write 1 table, free 1 table, repeat. Don't forget than in the whole sequence also the input data stays in shared-mem till the end.

Of course b) is only partially possible because we need to keep some tables to construct others.

**Comment by jgrosseo** [comment]
@pbuehler Do you want to comment? My impression was that the writer writes a table as soon as it arrives, and then clears the memory...

**Comment by pbuehler** [comment]
The aod-writer writes the tables as they arrive. However the aod-writer has no control of when the tables arrive or in which sequence. And the aod-writer does not delete tables after writing.

**Comment by jgrosseo** [comment]
@pbuehler Thanks. After that the tables are released by DPL then? @ktf?

**Comment by nburmaso** [comment]
Hi all. I did some tests to check the MC tracks filtration efficiency. When filtration is used, only reconstructed MC tracks and their relatives are saved into AOD. All primary particles are saved even if they are not reconstructed.

I used a script that Mihaela shared with me some time ago to estimate tree sizes. I used 100pbpb+100 signal events (1 TF) generated with the benchmarking script for this. The sample and the script can be found here https://cernbox.cern.ch/index.php/s/HmbSCYtZS3wTtPw

MC tracks filtration reduces the size of AOD significantly (see figure attached). Full size of the AOD file without filtration is ~5.7 GB, and with MC track filtration it is reduced down to ~37 MB. Excel tables with estimations, scripts and converted data can be found here https://cernbox.cern.ch/index.php/s/ZhMP1QK5CnuHw9h

I think with this we can just use AOD writer since the MC tracks table is not that huge if filtered. Perhaps, we can try optimizing MCKinematicsReader even further (if it is possible).

size_estim

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx [hunk 1]
```diff
@@ -217,6 +211,121 @@ void AODProducerWorkflowDPL::fillTracksTable(const TTracks& tracks, std::vector<
...
+    for (auto& mcTruth : mcTruthTPC) {
+      int source = mcTruth.getSourceID();
+      int event = mcTruth.getEventID();
+      int track = mcTruth.getTrackID();
+      toStore[source][event][track] = 1;
+    }
+  }
+  for (int source = 0; source < mcReader.getNSources(); source++) {
+    for (int event = 0; event < mcReader.getNEvents(source); event++) {
+      std::vector<MCTrack> const& mcParticles = mcReader.getTracks(source, event);
```
Comment (sawenzel) [inline Detectors/AOD/src/AODProducerWorkflowSpec.cxx:245]: there are 2 places that fetch the tracks. Are both of them needed?

Comment (nburmaso) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:245]: I think I can try fetching the tracks once and then use them in both `findRelatives()` and  `fillMCParticlesTable` . Thank you for the advice

**Comment by jgrosseo** [comment]
> 
> 
> Hi all. I did some tests to check the MC tracks filtration efficiency. When filtration is used, only reconstructed MC tracks and their relatives are saved into AOD. All primary particles are saved even if they are not reconstructed.
> 
> I used a script that Mihaela shared with me some time ago to estimate tree sizes. I used 100pbpb+100 signal events (1 TF) generated with the benchmarking script for this. The sample and the script can be found here https://cernbox.cern.ch/index.php/s/HmbSCYtZS3wTtPw
> 
> MC tracks filtration reduces the size of AOD significantly (see figure attached). Full size of the AOD file without filtration is ~5.7 GB, and with MC track filtration it is reduced down to ~37 MB. Excel tables with estimations, scripts and converted data can be found here https://cernbox.cern.ch/index.php/s/ZhMP1QK5CnuHw9h
> 
> I think with this we can just use AOD writer since the MC tracks table is not that huge if filtered. Perhaps, we can try optimizing MCKinematicsReader even further (if it is possible).
> 
> size_estim

Very interesting! Could you add the number of generated events and tracks to the spreadsheet, so that we can compare to the Run 2 conversion estimates?
At present we have there (for the full AOD, compressed size): PbPb: 1200kB/event and pp 24 kB/event

You may also check the logic Peter has implemented to reduce the MC particles which are kept. He presented this also here: https://indico.cern.ch/event/989361/contributions/4224741/attachments/2188557/3698406/AO2Dkine.pdf

**Comment by nburmaso** [comment]
Hi @jgrosseo. Sure, I will add the information on size per event/track soon

**Comment by nburmaso** [comment]
Hi @jgrosseo. I made new spreadsheets for size estimations. I generated 100 pbpb events with no signal embedding and 500 pp events. See figures attached. Converted data and the spreadsheets are available on CERNBox in `pbpb` and `pp` folders https://cernbox.cern.ch/index.php/s/ZhMP1QK5CnuHw9h

pbpb_size
pp_size

File: Detectors/AOD/src/AODProducerWorkflowSpec.cxx (hunk 1 above)
Comment (nburmaso) [reply Detectors/AOD/src/AODProducerWorkflowSpec.cxx:245]: Hi @sawenzel. I updated the code and merged `findRelatives()` and `fillMCParticlesTable()` into one function to fetch MC tracks only once (see latest commit)

**Comment by nburmaso** [comment]
> Hi @jgrosseo. I made new spreadsheets for size estimations. I generated 100 pbpb events with no signal embedding and 500 pp events. See figures attached. Converted data and the spreadsheets are available on CERNBox in `pbpb` and `pp` folders https://cernbox.cern.ch/index.php/s/ZhMP1QK5CnuHw9h
> 
> pbpb_size
> pp_size

Also note that there are events that were not reconstructed by the vertexer. There are 72 reconstructed vertices out of 100 PbPb events and 130 vertices out of 500 pp events. My estimations may be not quite correct due to this.

**Comment by jgrosseo** [comment]
> 
> 
> > Hi @jgrosseo. I made new spreadsheets for size estimations. I generated 100 pbpb events with no signal embedding and 500 pp events. See figures attached. Converted data and the spreadsheets are available on CERNBox in `pbpb` and `pp` folders https://cernbox.cern.ch/index.php/s/ZhMP1QK5CnuHw9h
> > pbpb_size
> > pp_size
> 
> Also note that there are events that were not reconstructed by the vertexer. There are 72 reconstructed vertices out of 100 PbPb events and 130 vertices out of 500 pp events. My estimations may be not quite correct due to this.

This is interesting. I see some large deviations from the run2 estimates. Would you have time to discuss these on Monday during the WP14? I fear that if we just show them on Wednesday at the plenary, we may not have enough time to digest them.

How do you calculate your average event size? With the number of reconstructed vertices?

For pp, I was surprised that: a) the bc table is larger than the mccollision table? It is 1000  times larger per collision than in Run2. How does this come about? b) why you don't have a mccollisionlabel table? This is needed for the reco-MC association.

In PbPb, there seems to be 7 times less tracks here than in the Run 2 MCs which we converted.

**Comment by jgrosseo** [comment]
image

**Comment by nburmaso** [comment]
Hi @jgrosseo.
>How do you calculate your average event size? With the number of reconstructed vertices?

I used number of generated events for these calculations. Perhaps, I should recalculate sizes per event with number of reconstructed vertices.

>the bc table is larger than the mccollision table? It is 1000 times larger per collision than in Run2. How does this come about?

I thought that it is due to continuous filling of the BC table. Isn't it so?

>why you don't have a mccollisionlabel table? This is needed for the reco-MC association

Honestly, I just missed it. I will add it as well. Do I understand correctly that we need to store MC eventID for each reconstructed vertex into `mccollisionlabel`?

>Would you have time to discuss these on Monday during the WP14? I fear that if we just show them on Wednesday at the plenary, we may not have enough time to digest them.

I will try to prepare something by Monday :)

**Comment by sawenzel** [comment]
**MEM CONSUMPTION NO FILTERING**

Screenshot 2021-03-11 at 10 56 45

**MEM CONSUMPTION WITH FILTERING**

Screenshot 2021-03-11 at 10 59 16

I made a quick memory study, comparing the version without filtering (screenshot on top) and the one with filtering (screenshot bottom). Each screenshot is showing the mem evolution (pss) of a) the source device, b) the producer, c) the AOD writer as well as the sum of all.

What can be seen is the following (This is for a study with 100PbPb + 100 signals in the dataframe): 

1. The original version needs ~40GB at an AOD output of ~4GB.
2. The filtered version still needs ~13GB at an AOD output of 4MB.
3. The input data (~1.6GB) from the source device (constant line) stays allocated throughout the whole lifetime.
4. (not shown): The CPU is 1-core at about 100% all the time (using --rate 1 to avoid spinning).

My conclusion is:

**The filtering has a nice effect and is a big improvement!**. 
However, 13GB is still a bit on the high side for a 1-CPU workflow. We should investigate why we still need ~10GB in the producer given the final output is only a couple of MB. Maybe there is still a problem in the track reader, or the tracks for one event are simply of that size, or some helper data structures take massive amounts of memory. 

I would suggest to merge this now, and keep looking for a second round of improvements.

For what the internal writing option is concerned: Let's keep it for now and potentially remove the workaround later.

**Comment by sawenzel** [comment]
@ktf: This is an example where https://alice.its.cern.ch/jira/browse/O2-1476 would really help. The constant line of the producer could be stopped when the data is fully consumed, freeing ~1.5GB for large parts of the lifetime.

**Pull request merged**